src/lib/homestore_backend/heap_chunk_selector.cpp

```diff
@@ -421,4 +438,88 @@ uint64_t HeapChunkSelector::total_blks(uint32_t dev_id) const {
     return it->second->m_total_blks;
 }
 
+std::list< uint32_t > HeapChunkSelector::get_pdev_ids() const {
+    std::list< uint32_t > pdev_ids;
+    for (const auto& [pdev_id, _] : m_per_dev_heap) {
+        pdev_ids.emplace_back(pdev_id);
+    }
+    return pdev_ids;
+}
+
+std::shared_ptr< HeapChunkSelector::ExtendedVChunk > HeapChunkSelector::select_empty_chunk_from_pdev(uint32_t pdev_id) {
+    std::unique_lock lock_guard(m_chunk_selector_mtx);
+    auto it = m_per_dev_heap.find(pdev_id);
+    if (it == m_per_dev_heap.end()) {
+        LOGWARNMOD(homeobject, "No pdev heap found for pdev {}", pdev_id);
+        return nullptr;
+    }
+    auto pdev_heap = it->second;
+    std::unique_lock lock(pdev_heap->mtx);
+    if (pdev_heap->m_heap.empty()) {
+        LOGWARNMOD(homeobject, "No available chunk found for pdev {}", pdev_id);
+        return nullptr;
+    }
+    auto chunk = pdev_heap->m_heap.top();
+    pdev_heap->m_heap.pop();
+    pdev_heap->available_blk_count -= chunk->available_blks();
+    RELEASE_ASSERT(chunk->available_blks() == chunk->get_total_blks(),
+                   "the selected chunk should be empty, chunk_id={}, pdev_id={}", chunk->get_chunk_id(), pdev_id);
+    RELEASE_ASSERT(!chunk->m_pg_id.has_value(), "chunk {} is selected from pdev heap, but it has a pg_id {}!",
+                   chunk->get_chunk_id(), chunk->m_pg_id.value());
+
+    return chunk;
+}
+
+std::shared_ptr< HeapChunkSelector::ExtendedVChunk >
+HeapChunkSelector::select_specific_chunk_from_pdev(uint32_t pdev_id, const chunk_num_t chunk_id) {
```

Comment:
The only user of this function is the recovery of reserve chunk.

We have similar requirement that exclude chunks belongs to PG from  adding into the per_dev_heap,  can we use similar approach , i.e changing `recover_per_dev_chunk_heap` to exclude reserve chunk as well?

Then we can kill this function

Reply:
good, point, will take a look

Reply:
I have remove those two function, ptal